_posts/2022-04-26-wsdl.markdown

```diff
@@ -0,0 +1,107 @@
+    {
+        serviceOptions.BaseAddresses.Add(new Uri("http://localhost/EchoService"));
+    })
+    .AddServiceEndpoint<EchoService, IEchoService>(new BasicHttpBinding(), "/basichttp");
+});
+```
+
+This will cause the service debug page to be available at the path /EchoService and the service endpoint to be available at /EchoService/basichttp. The `IServiceBuilder` interface also has a `BaseAddresses` property which would be applied to all configured services. The `ServiceOptions` class enables you to configure different base addresses for each service.
+
+You can also configure the listening address by setting `ServiceOptions.DebugBehavior.HttpHelpPageUrl`. If you want the service debug behavior to function for HTTPS requests, you need to have a service endpoint listening on HTTPS and then you can set `ServiceOptions.DebugBehavior.HttpsHelpPageEnabled` to true. There is a corresponding `HttpsHelpPageUrl` property to override the Url for the HTTPS help page.
+
+If you are following along testing this feature, you may have noticed that the help page looks identical to the WCF help page and provides the help content that you would expect for configuring WCF on .NET Framework. This may seem unusual, but this is for compatibility reasons. We received very specific feedback that some companies use the help page for automatic health check systems and look for specific text to appear on the page. If a service has failed to start for some reason, WCF hosted in IIS/ASP.NET will still serve an html page with different content explaining the failed service startup. Keeping the help page text identical to WCF enables switching to CoreWCF without the need to modify any existing infrastucture which depends on this page as a basic service health check.
+
+#### ServiceMetadataBehavior
+
+To enable exporting the service WSDL, in addition to calling `AddServiceModelMetadata()`, you also need to enable getting the WSDL. Calling `AddServiceModelMetadata()` adds the behavior `ServiceMetadataBehavior` to DI as a Singleton which automatically gets picked up by CoreWCF when creating the service dispatcher. To make any changes to `ServiceMetadataBehavior`, you can request it from DI and make any changes. At it's simplest, you can add the following code to your service configuration:
+
+```csharp
+var serviceMetadataBehavior = app.Services.GetRequiredService<CoreWCF.Description.ServiceMetadataBehavior>();
+serviceMetadataBehavior.HttpGetEnabled = true;
+```
+
+This enables the WSDL to be served via HTTP. Service metadata has the same behavior for its listening Url that service debug behavior has. The same or similar mechanisms work to change which address the WSDL will be available from. There is one more change you are likely to want to make to have CoreWCF modify the generated help page and WSDL document to reflect the hostname and port number that the request came from. This can be done by adding `UseRequestHeadersForMetadataAddressBehavior` to DI like this:
```

Comment:
Why isn't this the default?

Reply:
Because it's only the default in WCF when you host in IIS and you enable the UseMultipleSiteBindings feature. There's actually a small perf hit to using this feature as it has to re-write the WSDL on every request.